invenio_communities/notifications/builders.py

```diff
@@ -0,0 +1,59 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright (C) 2023 Graz University of Technology.
+#
+# Invenio-Communities is free software; you can redistribute it and/or modify
+# it under the terms of the MIT License; see LICENSE file for more details.
+
+"""Notification related utils for notifications."""
+
+from invenio_notifications.models import Notification
+from invenio_notifications.registry import EntityResolverRegistry
+from invenio_notifications.services.builders import NotificationBuilder
+from invenio_notifications.services.generators import EntityResolve, UserEmailBackend
+from invenio_users_resources.notifications.filters import UserPreferencesRecipientFilter
+from invenio_users_resources.notifications.generators import UserRecipient
+
+
+class CommunityInvitationNotificationBuilder(NotificationBuilder):
+    """Base notification builder for community invitation events."""
+
+    type = "community-invitation"
+
+    @classmethod
+    def build(cls, request, role, message=None):
+        """Build notification with request context."""
+        return Notification(
+            type=cls.type,
+            context={
+                "request": EntityResolverRegistry.reference_entity(request),
+                "role": role,
+                "message": message,
+            },
+        )
+
+    context = [
+        EntityResolve(key="request"),
+        EntityResolve(key="request.created_by"),
+        EntityResolve(key="request.receiver"),
+    ]
+
+    recipients = [
+        UserRecipient(key="request.receiver"),
+    ]
+
+    recipient_filters = [
+        UserPreferencesRecipientFilter(),
+    ]
+
+    recipient_backends = [
+        UserEmailBackend(),
+    ]
+
+
+class CommunityInvitationSubmittedNotificationBuilder(
+    CommunityInvitationNotificationBuilder
```

Comment:
minor: do we want to create these base classes always? E.g. in this case what other notification would inherit from this class?

Reply:
Accept/Decline invitation notifications could inherit from the base class as well. In both cases, recipients (certain community members should probably be informed (owners, managers?)) and the build method will have to be updated (no need to include role and message).

Reply:
When implementing accept/decline, we can refactor the base class to only include parts shared by all notifications.

Reply:
+1 for simplifying the subclassing until needed